In [ ]:
# these are the libraries which we would use but pandas is used very rare

import pandas as pd
import numpy as np

 **To explore the the data :**

---



 *The data set is included in scikit-learn in database module*




In [ ]:
#we can load it by calling load_iris function

from sklearn.datasets import load_iris
iris_dataset = load_iris()

*This iris return a bunch object , which is similar to dictionary(contains keys and values)*


In [ ]:
#the following code will show all the keys(its a term referred with dictionary) of the dataset which we have used in this project

print("Keys of iris_dataset: \n{}".format(iris_dataset.keys()))

Keys of iris_dataset: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


**Now lets see every keys which are included in the data set :**

1.  *First we will start with the* **DESCR** : *this key provides the description of the datasets*



In [ ]:
# here we are printing only 200 characteristics (193 in video session) using the slicing method

print(iris_dataset['DESCR'][:200] + "\n..." )
# this is a one line code

#the above code can be written in begginer freindly way

#val = iris_dataset['DESCR']
#start_val = val[:200]
#print( start_val + "\n...")

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive
...


2. *We will proceed to* **target_names** : *contains the species of flower that we want to predict i.e. 'setosa' 'versicolor' and 'virginica'*

In [ ]:
# printing the target names with format feature of python

print("Target names: {}".format(iris_dataset['target_names']))

Target names: ['setosa' 'versicolor' 'virginica']


3. *Now we are heading with* **feature_names** : *gives the description of each feature it includes the 'sepal length', 'sepal width', 'petal length' and the 'petal width' all in centimeters.* 

In [ ]:
# again using format specification of python

print("Feature names: \n{}".format(iris_dataset['feature_names'])) 

Feature names: 
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


4. *Heading ahead with* **data** : *contains numeric measurements of sepal length, sepal width, petal length, and petal width in a NumPy array*




In [ ]:
# this will print the type of key-"data'

print( "Type of data : {}".format(type(iris_dataset['data'])))

Type of data : <class 'numpy.ndarray'>


*The shape of the data array is the number of samples multiplied by the number of features. This is a convention in scikit-learn, and your data will always be assumed to be in this shape*

In [ ]:
#this will print the shape (whether it's a 1-day array or 2-d or any other) of the data

print("Shape of the data: {}".format(iris_dataset['data'].shape))

Shape of the data: (150, 4)


-------**data** key is a **two-dimensional array**------

The rows in this data array corrrespond to the flowers, while the columns represent the four measurements that were taken for each flower

*Here are the feature values for the first five samples:*

In [ ]:
# printing will be done using slicing way to print 5 samples from 150 samples

print("First five columns of data: \n{}".format(iris_dataset['data'][:5]))

First five columns of data: 
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


5. Now lets come to **target** key : contains the species of each flowers that were measured

In [ ]:
# this will print the type of key-"data'

print( "Type of target : {}".format(type(iris_dataset['target'])))

Type of target : <class 'numpy.ndarray'>


**target** is a **one-dimensional array**, and on seeing the **shape** we can see that it contains **one entry per flower**

In [ ]:
#this will print the shape  of the target

print("Shape of the target: {}".format(iris_dataset['target'].shape))

Shape of the target: (150,)


On seeing the target key and exploring the values we can easily see that The Species are enclosed as integers from 0 to 2:

In [ ]:
# print the values of "target" key
print("Target: \n{}".format(iris_dataset['target']))

Target: 
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


*The meanings of above values :* 

a)  **0** means **setosa**

b)  **1** means **versicolor**

c)  **2** means **virginica** 


**TRAIN AND TESTING OF MODEL**

---



We will now first train the model and test if its working fine or not

*We are  going to import scikit-learn to, here in the below code :-*

X(capital) - Denoted for **data** / Its two dimensional array (a matrix)

y(lowercase) - denotes **labels** / Its one- dimensional array(a vector)

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(iris_dataset['data'], iris_dataset['target'], random_state=0)

*Before making the split, the train_test_split function shuffles the dataset using a pseudorandom number generator. To make sure that we will get the same output if we run the same function several times, we provide the pseudorandom number generator with a fixed seed using the random_state parameter.*

The output of the train_test_split function is X_test,X_train,y_test and y_train, which all are NumPy arrays. X_train contains 75% of the rows of the dataset, abd X_test contains the remaining 25%. We will see all the piece of this paragraph with the following code:

In [ ]:
print("X_train shape: {}".format(X_train.shape))
print("y_train shape: {}".format(y_train.shape))
print("X_test shape: {}".format(X_test.shape))
print("y_test shape: {}".format(y_test.shape))

X_train shape: (112, 4)
y_train shape: (112,)
X_test shape: (38, 4)
y_test shape: (38,)


 ***INSPECTING THE DATA***

---



Now we start the real machine learning model. This model can be solved with many classification model but we are using the k-nearest neighbours classifier.

**k-nearest neighbours classifier : ** *In this clssification algorithm, to make prediction of a new data point, the classifier finds the point in the training set that is closest to the new point. Then this assigns the label of this training point to the new data point.*

In [ ]:
#The most important parameter of KneighboursClassifier is the number of neighbours, which we will set 1
# this code is to import KNeighborsClassifier
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=1)

knn will hold the information that the algorithm has extracted from the training data. The KneighborsClassification will only store the information 

*To build the model on the training set we call the fit method of the knn object :*

In [ ]:
# knn object,takes as argument the NumPy array X_train containing the training sata and the NumPy array y_train of the corresponding training 
#labels:
# This code is training the model 
knn.fit(X_train, y_train)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

**MAKING PREDICTIONS**

---
We have trained the model now we are gonna predict that the model is running good or not


In [ ]:
# Here we are providing an unknown flowers featuers that are : sepal_lenth=5cm,sepal width=2.9cm petal length=1cm,petal width=0.2cm
# we will put these data into a numPy array, calculating the shape (no_of_samples(1) x no_of_features(4))

X_new = np.array([[5, 2.9, 1, 0.2]])
print("X_new shape: {}".format(X_new.shape))

# we accept the argument of np.array as two dimensional because scikit-learn always expects two dimensional arrrays of data

X_new shape: (1, 4)


*To make prediction we call the predict method of the knn object*

In [ ]:
prediction = knn.predict(X_new)

print("Prediction: {}".format(prediction))
print("Predicted target name: {}".format(iris_dataset['target_names'][prediction]))

Prediction: [0]
Predicted target name: ['setosa']


**EVALUATING THE MODEL**

---
We can now measure how well th emodel works by computing the accuracy, which is the fraction of flowers for which the right species was predicted:


In [ ]:
y_pred=knn.predict(X_test)
print("Test set predictions :\n {}".format(y_pred))

Test set predictions :
 [2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


In [ ]:
# now we will check if the predicted value is same as that of assigned values
print("Test set score : {}".format(np.mean(y_pred == y_test)))

Test set score : 0.9736842105263158


*So, At last we got our **model** with aprroximately **97% accuracy** which is a appreciable work*